### PR TITLE
Swift: flow through writeable keypaths

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -476,6 +558,20 @@ private module ArgumentNodes {
...
+    private KeyPathApplicationExprCfgNode keyPath;
+
+    KeyPathAssignmentArgumentNode() {
+      keyPath = this.getCfgNode() and
+      exists(AssignExpr assign | assign.getDest() = keyPath.getNode().asAstNode())
+    }
+
+    override predicate argumentOf(DataFlowCall call, ArgumentPosition pos) {
+      call.asKeyPath() = keyPath and
+      pos = TPositionalArgument(0) // TODO: new parameter type?
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:572]: Same here - I think this is fine as-is.

**Review by MathiasVP — COMMENTED** [review]
A couple of comments, but generally this LGTM. We should also make sure to run DCA on this before we merge it.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -735,6 +831,24 @@ predicate storeStep(Node node1, ContentSet c, Node node2) {
...
+    )
+  |
+    // the previous node is either the next element in the chain
+    node1.(KeyPathComponentPostUpdateNode).getComponent() = component.getNextComponent()
+    or
+    // or the return node, if this is the last component in the chain
+    not exists(component.getNextComponent()) and
+    node1.(KeyPathReturnPostUpdateNode).getKeyPathExpr() = component.getKeyPathExpr()
+  )
+  or
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:851]: We don't seem to... Is there an easy way to check what reverse reads are being generated>

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -227,6 +248,17 @@ private module Cached {
...
+    nodeFrom.(KeyPathComponentPostUpdateNode).getComponent() = 
+      nodeTo.(KeyPathParameterPostUpdateNode).getComponent(0)
+    or
+    // Flow to the result of a keypath assignment
+    // TODO: is there a cleaner way to do this?
+    exists(KeyPathApplicationExpr apply, AssignExpr assign |
+      apply = assign.getDest() and
+      nodeTo.asExpr() = apply and
+      nodeFrom.asExpr() = assign.getSource()
+    )
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:260]: that doesn't quite seem to do what I want here - the current implementation steps from the source to the preupdate node, then there's interprocedural flow through the keypath to the postupdate node. With the WriteDefinition that first step goes to the WriteDefinition and then skips the postupdate node entirely, from what I can tell....

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:851]: You can quick-eval `store` in `storeSet` in `DataFlowImplCommon.qll` (the reverse read case is the disjunct with `readSet(n2, c, n1)` in it). There's also a consistency query called `reverseRead` for it.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:264]: Oh, I see. Yeah, I can see how this is different. Maybe this is simply how we need to do things then 🤔

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: Figured it out - I hadn't made KeyPathComponentPostUpdateNode a subclass of PostUpdateNodeImpl

**Review by MathiasVP — COMMENTED** [review]
LGTM (except for the formatting issue) 👍

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**